client-spark/spark2/src/main/java/org/apache/spark/shuffle/RssShuffleManager.java

```diff
@@ -446,6 +442,7 @@ public <K, V> ShuffleWriter<K, V> getWriter(
           this,
           sparkConf,
           shuffleWriteClient,
+          this::getOrCreateShuffleManagerClient,
```

Comment:
I think the lifecycle of `ShuffleManagerClient` may be different with the shuffle manager, if this is idle with some times, it should be closed(this is not supported now). If that, we should provide the wrapper to create or get the active `ShuffleManagerClient` .

Reply:
> I think the lifecycle of `ShuffleManagerClient` may be different with the shuffle manager, if this is idle with some times, it should be closed(this is not supported now). If that, we should provide the wrapper to create or get the active `ShuffleManagerClient` .

agree with u. the lifecycle of `ShuffleManagerClient` is major factor in guiding how to modify

Reply:
> hmm, maybe I wasn't clear enough and mislead you. It's fine and probably better to accept a `ShuffleManagerClient` in builders instead of a supplier.
> 
> Instead of passing the `shuffleManagerClient` directly, we probably want to call `getOrCreateShuffleManagerClient()` instead to make sure the client is actually initialized.

Do you might be considering make `ShuffleManagerClient` as a single object or a `LazyShuffleManagerClient`  impl which logic like a `lazy val shuffleManagerClient` in Scala ?

Reply:
Properly not. Maybe like the following  code.

```java
class AutoClosableShuffleManagerClient extends ShuffleManagerClient {
    private ShuffleManagerClient client;
    private int reference = 0

    public get() {
      reference += 1;
      if (client == null) {
        client = new ShuffleManagerClient();
      }
    }

    public void close() {
      reference -= 1

      if (reference == 0) {
        client.close()
        client = null
      }
    }
  }
```

Maybe this is not right, just my simple thought.